File: src/core/HistoryManager.ts
import * as path from 'node:path';
import * as fs from 'node:fs/promises';
import { existsSync } from 'node:fs';
import { AudioService } from '../infrastructure/AudioService.js';
import { ISettingsManager } from './interfaces/ISettingsManager.js';

/**
 * Manages the synthesized audio history.
 * Handles the persistence of metadata inside a JSON file and rotates/renames
 * corresponding WAV files on the local filesystem to enforce storage caps.
 */
export class HistoryManager {
    /** Target directory where history files are saved */
    private audioDir: string;
    
    /** Absolute path to the JSON file tracking history items */
    private historyPath: string;
    
    /** Maximum history items kept on disk before rotation overrides old records */
    private readonly MAX_HISTORY = 20;

    constructor(
        private audioService: AudioService,
        private settingsManager: ISettingsManager
    ) {
        this.audioDir = path.join(settingsManager.getAppDirectory(), 'audio');
        this.historyPath = path.join(this.audioDir, 'history.json');
    }

    /**
     * Adds a newly synthesized speech entry into history.
     * Enforces size limits by shifting existing files (`tts_output_i.wav` to `tts_output_i+1.wav`)
     * and writing the text metadata to `history.json`.
     * @param text Original synthesized sentence.
     * @param tempWavPath Path to temporary generated raw WAV file.
     */
    public async addEntry(text: string, tempWavPath: string): Promise<void> {
        await fs.mkdir(this.audioDir, { recursive: true });

        // Rotate history files starting from the second to last index down to zero.
        // E.g., shifts tts_output_18.wav -> tts_output_19.wav, ..., tts_output_0.wav -> tts_output_1.wav
        for (let i = this.MAX_HISTORY - 2; i >= 0; i--) {
            const oldFile = path.join(this.audioDir, `tts_output_${i}.wav`);
            const newFile = path.join(this.audioDir, `tts_output_${i + 1}.wav`);
            try {
                await fs.rename(oldFile, newFile);
            } catch (err: any) {
                // Ignore missing file errors (ENOENT) during initial runs or if history is sparse
                if (err.code !== 'ENOENT') {
                    console.error(`Failed to rotate history file ${oldFile}:`, err.message);
                }
            }
        }

        // Copy the new audio to index 0
        const newFilePath = path.join(this.audioDir, `tts_output_0.wav`);
        await fs.copyFile(tempWavPath, newFilePath);

        // Update history.json list
        let historyTexts: string[] = [];
        try {
            if (existsSync(this.historyPath)) {
                const historyData = await fs.readFile(this.historyPath, 'utf-8');
                historyTexts = JSON.parse(historyData);
            }
        } catch (e) {
            // Suppress errors and fallback to empty history if parsing fails
        }
        
        historyTexts.unshift(text);
        if (historyTexts.length > this.MAX_HISTORY) {
            historyTexts = historyTexts.slice(0, this.MAX_HISTORY);
        }
        await fs.writeFile(this.historyPath, JSON.stringify(historyTexts, null, 2), 'utf-8');
    }

    /**
     * Retrieves the history metadata from disk.
     * @returns List of synthesized sentences in order of newest to oldest.
     */
    public async getHistory(): Promise<string[]> {
        if (!existsSync(this.historyPath)) {
            return [];
        }
        try {
            const data = await fs.readFile(this.historyPath, 'utf-8');
            return JSON.parse(data);
        } catch (e) {
            return [];
        }
    }

    /**
     * Plays a historical recording by its unique index ID.
     * Routes the call to AudioService with current playback configurations.
     * @param id The index of the item (0 is newest).
     */
    public async playHistory(id: number): Promise<void> {
        const filePath = path.join(this.audioDir, `tts_output_${id}.wav`);
        if (!existsSync(filePath)) {
            console.log(`Audio file for ID [${id}] not found.`);
            return;
        }

        try {
            const appConfig = await this.settingsManager.getAppConfig();
            console.log(`Playing history [${id}]...`);
            await this.audioService.play(
                filePath, 
                appConfig.playback,
                appConfig.playbackDevice, 
                appConfig.volume, 
                appConfig.monitoring, 
                appConfig.monitoringDevice, 
                appConfig.monitoringVolume
            );
        } catch (error: any) {
            console.error('Failed to play history audio:', error.message);
        }
    }

    /**
     * Deletes a history entry, shifts subsequent files down to close the file index gap,
     * and updates the metadata manifest.
     * @param id Index of the history entry to delete.
     */
    public async deleteHistory(id: number): Promise<void> {
        let historyTexts = await this.getHistory();
        
        if (id >= historyTexts.length || id < 0) {
            console.log(`History ID [${id}] not found.`);
            return;
        }

        // Remove the target entry from metadata
        historyTexts.splice(id, 1);
        await fs.writeFile(this.historyPath, JSON.stringify(historyTexts, null, 2), 'utf-8');

        // Delete the WAV file
        const fileToDelete = path.join(this.audioDir, `tts_output_${id}.wav`);
        if (existsSync(fileToDelete)) {
            await fs.unlink(fileToDelete);
        }

        // Shift all subsequent files down (e.g. tts_output_3.wav renames to tts_output_2.wav)
        for (let i = id + 1; i < this.MAX_HISTORY; i++) {
            const oldFile = path.join(this.audioDir, `tts_output_${i}.wav`);
            const newFile = path.join(this.audioDir, `tts_output_${i - 1}.wav`);
            if (existsSync(oldFile)) {
                await fs.rename(oldFile, newFile);
            }
        }
        console.log(`Deleted history [${id}] and shifted remaining IDs.`);
    }

    /**
     * Deletes all synthesized history files and removes the metadata manifest.
     */
    public async clearHistory(): Promise<void> {
        if (!existsSync(this.audioDir)) {
            console.log('No history to clear.');
            return;
        }
        try {
            const files = await fs.readdir(this.audioDir);
            for (const file of files) {
                if ((file.startsWith('tts_output_') && file.endsWith('.wav')) || file === 'history.json') {
                    await fs.unlink(path.join(this.audioDir, file));
                }
            }
            console.log('History cleared.');
        } catch (error: any) {
            console.error('Failed to clear history:', error.message);
        }
    }

    /**
     * Prints history logs to stdout.
     */
    public async printHistory(): Promise<void> {
        const historyTexts = await this.getHistory();

        if (historyTexts.length === 0) {
            console.log('No history available yet.');
            return;
        }

        console.log('Recent TTS Audio History (Newest to Oldest):');
        historyTexts.forEach((text, index) => {
            let displayText = text.replace(/\\n/g, ' ').trim();
            if (displayText.length > 50) {
                displayText = displayText.substring(0, 50) + '...';
            }
            console.log(`[${index}] ${displayText}`);
        });
    }
}

File: src/core/history_manager.py
"""History manager for Moon-TTS application.

Maintains a rolling list of the most recent TTS outputs as numbered WAV
files and a parallel ``history.json`` metadata file.
"""

from __future__ import annotations

import json
import os
import shutil
import threading
from pathlib import Path
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from src.core.settings_manager import SettingsManager


class HistoryManager:
    """Manages a capped history of TTS audio outputs.

    Audio files are stored as ``tts_output_{uuid}.wav``.
    Adding a new entry writes the file and records it in ``history.json``
    with a unique ID and text. Oldest entries are deleted on overflow.
    """

    MAX_HISTORY: int = 20

    def __init__(self, audio_service: Any, settings_manager: "SettingsManager") -> None:
        """Initialize the history manager.

        Args:
            audio_service: Service used to play back audio files.
            settings_manager: Application settings manager instance.
        """
        self._audio_service = audio_service
        self._settings_manager = settings_manager

        app_dir = settings_manager.get_app_directory()
        self._audio_dir: Path = Path(app_dir) / "audio"
        self._history_path: Path = self._audio_dir / "history.json"
        self._lock = threading.RLock()

        os.makedirs(self._audio_dir, exist_ok=True)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def add_entry(self, text: str, temp_wav_path: str) -> None:
        """Add a new TTS output to the history.

        Saves the WAV file using a unique ID, prepends it to the history metadata,
        and deletes any oldest files exceeding MAX_HISTORY.

        Args:
            text: The text that was synthesised.
            temp_wav_path: Path to the temporary WAV file to archive.
        """
        import uuid
        with self._lock:
            history = self._load_history()

            # Generate unique ID and target path
            entry_id = str(uuid.uuid4())
            dst_path = self._wav_path(entry_id)

            # Copy temp file to target
            shutil.copy2(temp_wav_path, str(dst_path))

            # Prepend new entry
            history.insert(0, {"id": entry_id, "text": text})

            # Check for overflow and delete files
            if len(history) > self.MAX_HISTORY:
                to_delete = history[self.MAX_HISTORY:]
                history = history[: self.MAX_HISTORY]
                for item in to_delete:
                    old_path = self._wav_path(item["id"])
                    if os.path.exists(old_path):
                        try:
                            os.remove(old_path)
                        except Exception:
                            pass

            self._save_history(history)

    def get_history(self) -> list[dict[str, Any]]:
        """Return the list of history text entries.

        Returns:
            List of dicts containing 'id' and 'text', newest first.
        """
        with self._lock:
            return list(self._load_history())

    def play_history(self, entry_id: str) -> None:
        """Play a historical TTS output by its unique ID.

        Args:
            entry_id: Unique string ID of the history entry.
        """
        with self._lock:
            if self._audio_service.is_playing():
                return
            wav = self._wav_path(entry_id)
            if not os.path.exists(wav):
                return

            config = self._settings_manager.get_app_config()
            self._audio_service.play_with_config(str(wav), config)

    def delete_history(self, entry_id: str) -> None:
        """Delete a single history entry by its unique ID.

        Args:
            entry_id: Unique string ID of the entry to remove.
        """
        with self._lock:
            history = self._load_history()

            # Remove the WAV file
            target = self._wav_path(entry_id)
            if os.path.exists(target):
                try:
                    os.remove(target)
                except Exception:
                    pass

            # Filter history list
            new_history = [item for item in history if item["id"] != entry_id]
            self._save_history(new_history)

    def clear_history(self) -> None:
        """Delete all history audio files and reset the metadata."""
        with self._lock:
            history = self._load_history()
            for item in history:
                wav = self._wav_path(item["id"])
                if os.path.exists(wav):
                    try:
                        os.remove(wav)
                    except Exception:
                        pass

            self._save_history([])

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _wav_path(self, entry_id: str) -> Path:
        """Return the path for the WAV file with the given entry_id.

        Args:
            entry_id: Unique ID of the entry.

        Returns:
            Full path to ``tts_output_{entry_id}.wav``.
        """
        return self._audio_dir / f"tts_output_{entry_id}.wav"

    def _load_history(self) -> list[dict[str, Any]]:
        """Read history.json and return its contents.

        Returns:
            List of dictionary entries, or empty list on error / missing file.
        """
        with self._lock:
            if not self._history_path.exists():
                return []
            try:
                with open(self._history_path, "r", encoding="utf-8") as fh:
                    data = json.load(fh)
                    if not isinstance(data, list):
                        return []
                    cleaned = []
                    for item in data:
                        if isinstance(item, dict) and "id" in item and "text" in item:
                            cleaned.append(item)
                    return cleaned
            except (json.JSONDecodeError, ValueError):
                return []

    def _save_history(self, history: list[dict[str, Any]]) -> None:
        """Write *history* to history.json.

        Args:
            history: List of dictionary entries to persist.
        """
        with self._lock:
            os.makedirs(self._audio_dir, exist_ok=True)
            with open(self._history_path, "w", encoding="utf-8") as fh:
                json.dump(history, fh, indent=2, ensure_ascii=False)
